File: test/l2/l2_dai_bridge.py
import pytest

from starkware.starknet.testing.starknet import Starknet
from starkware.starknet.testing.contract import StarknetContract
from starkware.starkware_utils.error_handling import StarkException
from starkware.starknet.business_logic.execution.objects import Event
from starkware.starknet.public.abi import get_selector_from_name
from itertools import chain

L1_ADDRESS = 0x1
INVALID_L1_ADDRESS = 0x10000000000000000000000000000000000000000
L1_BRIDGE_ADDRESS = 0x1
FINALIZE_WITHDRAW = 0
ECDSA_PUBLIC_KEY = 0

burn = 0
no_funds = 1

starknet_contract_address = 0x0


###########
# HELPERS #
###########
def to_split_uint(a):
    return (a & ((1 << 128) - 1), a >> 128)

def to_uint(a):
    return a[0] + (a[1] << 128)

def check_event(contract, event_name, tx, values):
    expected_event = Event(
        from_address=contract.contract_address,
        keys=[get_selector_from_name(event_name)],
        data=list(chain(*[e if isinstance(e, tuple) else [e] for e in values]))
    )
    assert expected_event in ( tx.raw_events if hasattr(tx, 'raw_events') else tx.get_sorted_events())

#########
# TESTS #
#########
@pytest.mark.asyncio
async def test_initiate_withdraw(
    starknet: Starknet,
    dai: StarknetContract,
    l2_bridge: StarknetContract,
    user1: StarknetContract,
    check_balances,
):
    await dai.approve(
            l2_bridge.contract_address,
            to_split_uint(10),
        ).invoke(user1.contract_address)

    tx = await l2_bridge.initiate_withdraw(
            L1_ADDRESS,
            to_split_uint(10)).invoke(user1.contract_address)

    check_event(
        l2_bridge,
        'withdraw_initiated',
        tx,
        (L1_ADDRESS, to_split_uint(10), user1.contract_address)
    )

    payload = [FINALIZE_WITHDRAW, L1_ADDRESS, *to_split_uint(10)]
    starknet.consume_message_from_l2(
        from_address=l2_bridge.contract_address,
        to_address=L1_BRIDGE_ADDRESS,
        payload=payload,
    )

    await check_balances(90, 100)


@pytest.mark.asyncio
async def test_close_should_fail_when_not_authorized(
    l2_bridge: StarknetContract,
    user1: StarknetContract,
):
    with pytest.raises(StarkException) as err:
        await l2_bridge.close().invoke(user1.contract_address)
    assert "l2_dai_bridge/not-authorized" in str(err.value)


@pytest.mark.asyncio
async def test_initiate_withdraw_should_fail_when_closed(
    starknet: Starknet,
    dai: StarknetContract,
    l2_bridge: StarknetContract,
    auth_user: StarknetContract,
    user1: StarknetContract,
    user2: StarknetContract,
):
    await dai.approve(
            l2_bridge.contract_address,
            to_split_uint(10),
        ).invoke(user1.contract_address)

    await l2_bridge.close().invoke(auth_user.contract_address)

    with pytest.raises(StarkException) as err:
        await l2_bridge.initiate_withdraw(
                user2.contract_address,
                to_split_uint(10)).invoke(user1.contract_address)
    assert "l2_dai_bridge/bridge-closed" in str(err.value)

    with pytest.raises(AssertionError):
        payload = [FINALIZE_WITHDRAW, L1_ADDRESS, *to_split_uint(10)]
        starknet.consume_message_from_l2(
            from_address=l2_bridge.contract_address,
            to_address=L1_BRIDGE_ADDRESS,
            payload=payload,
        )


@pytest.mark.asyncio
async def test_initiate_withdraw_insufficient_funds(
    starknet: Starknet,
    l2_bridge: StarknetContract,
    user3: StarknetContract,
):
    with pytest.raises(StarkException) as err:
        await l2_bridge.initiate_withdraw(
                L1_ADDRESS,
                to_split_uint(10)).invoke(user3.contract_address)
    assert "dai/insufficient-balance" in str(err.value)

    with pytest.raises(AssertionError):
        payload = [FINALIZE_WITHDRAW, L1_ADDRESS, *to_split_uint(10)]
        starknet.consume_message_from_l2(
            from_address=l2_bridge.contract_address,
            to_address=L1_BRIDGE_ADDRESS,
            payload=payload,
        )


@pytest.mark.asyncio
async def test_withdraw_invalid_l1_address(
    starknet: Starknet,
    dai: StarknetContract,
    l2_bridge: StarknetContract,
    user1: StarknetContract,
    check_balances,
):
    await dai.approve(
            l2_bridge.contract_address,
            to_split_uint(10),
        ).invoke(user1.contract_address)
    with pytest.raises(StarkException) as err:
        await l2_bridge.initiate_withdraw(
                INVALID_L1_ADDRESS,
                to_split_uint(10)).invoke(user1.contract_address)
    assert "l2_dai_bridge/invalid-l1-address" in str(err.value)

    payload = [FINALIZE_WITHDRAW, INVALID_L1_ADDRESS, *to_split_uint(10)]
    with pytest.raises(AssertionError):
        starknet.consume_message_from_l2(
            from_address=l2_bridge.contract_address,
            to_address=L1_BRIDGE_ADDRESS,
            payload=payload,
        )

    await check_balances(100, 100)


@pytest.mark.asyncio
async def test_handle_deposit(
    starknet: Starknet,
    l2_bridge: StarknetContract,
    user2: StarknetContract,
    check_balances,
):
    tx = await starknet.send_message_to_l2(
        from_address=L1_BRIDGE_ADDRESS,
        to_address=l2_bridge.contract_address,
        selector="handle_deposit",
        payload=[
            user2.contract_address,
            *to_split_uint(10),
            L1_ADDRESS
        ],
    )

    check_event(
        l2_bridge, 'deposit_handled', tx, (user2.contract_address, to_split_uint(10))
    )

    await check_balances(100, 110)


@pytest.mark.asyncio
async def test_handle_force_withdrawal(
    starknet: Starknet,
    dai: StarknetContract,
    l2_bridge: StarknetContract,
    user1: StarknetContract,
    check_balances,
):
    await dai.approve(
            l2_bridge.contract_address,
            to_split_uint(10),
        ).invoke(user1.contract_address)

    tx = await starknet.send_message_to_l2(
        from_address=L1_BRIDGE_ADDRESS,
        to_address=l2_bridge.contract_address,
        selector="handle_force_withdrawal",
        payload=[
            user1.contract_address,
            int(L1_ADDRESS),
            *to_split_uint(10)
        ],
    )

    check_event(
        l2_bridge,
        'force_withdrawal_handled',
        tx,
        (int(L1_ADDRESS), to_split_uint(10), user1.contract_address)
    )

    payload = [FINALIZE_WITHDRAW, L1_ADDRESS, *to_split_uint(10)]
    starknet.consume_message_from_l2(
        from_address=l2_bridge.contract_address,
        to_address=L1_BRIDGE_ADDRESS,
        payload=payload,
    )

    await check_balances(90, 100)


@pytest.mark.asyncio
async def test_handle_force_withdrawal_insufficient_funds(
    starknet: Starknet,
    dai: StarknetContract,
    l2_bridge: StarknetContract,
    user3: StarknetContract,
):
    await dai.approve(
            l2_bridge.contract_address,
            to_split_uint(10),
        ).invoke(user3.contract_address)

    tx = await starknet.send_message_to_l2(
        from_address=L1_BRIDGE_ADDRESS,
        to_address=l2_bridge.contract_address,
        selector="handle_force_withdrawal",
        payload=[
            user3.contract_address,
            int(L1_ADDRESS),
            *to_split_uint(10)
        ],
    )

    check_event(
        l2_bridge,
        'force_withdrawal_handled',
        tx,
        (int(L1_ADDRESS), to_split_uint(10), user3.contract_address)
    )

    with pytest.raises(AssertionError):
        payload = [FINALIZE_WITHDRAW, L1_ADDRESS, *to_split_uint(10)]
        starknet.consume_message_from_l2(
            from_address=l2_bridge.contract_address,
            to_address=L1_BRIDGE_ADDRESS,
            payload=payload,
        )


@pytest.mark.asyncio
async def test_handle_force_withdrawal_insufficient_allowance(
    starknet: Starknet,
    l2_bridge: StarknetContract,
    user1: StarknetContract,
):
    tx = await starknet.send_message_to_l2(
        from_address=L1_BRIDGE_ADDRESS,
        to_address=l2_bridge.contract_address,
        selector="handle_force_withdrawal",
        payload=[
            user1.contract_address,
            int(L1_ADDRESS),
            *to_split_uint(10)
        ],
    )

    check_event(
        l2_bridge,
        'force_withdrawal_handled',
        tx,
        (int(L1_ADDRESS), to_split_uint(10), user1.contract_address)
    )

    with pytest.raises(AssertionError):
        payload = [FINALIZE_WITHDRAW, L1_ADDRESS, *to_split_uint(10)]
        starknet.consume_message_from_l2(
            from_address=l2_bridge.contract_address,
            to_address=L1_BRIDGE_ADDRESS,
            payload=payload,
        )


@pytest.mark.asyncio
async def test_handle_force_withdrawal_invalid_l1_address(
    starknet: Starknet,
    dai: StarknetContract,
    l2_bridge: StarknetContract,
    user1: StarknetContract,
    check_balances,
):
    await dai.approve(
            l2_bridge.contract_address,
            to_split_uint(10),
        ).invoke(user1.contract_address)

    tx = await starknet.send_message_to_l2(
        from_address=L1_BRIDGE_ADDRESS,
        to_address=l2_bridge.contract_address,
        selector="handle_force_withdrawal",
        payload=[
            user1.contract_address,
            int(INVALID_L1_ADDRESS),
            *to_split_uint(10)
        ],
    )

    check_event(
        l2_bridge,
        'force_withdrawal_handled',
        tx,
        (int(INVALID_L1_ADDRESS), to_split_uint(10), user1.contract_address)
    )

    with pytest.raises(AssertionError):
        payload = [FINALIZE_WITHDRAW, INVALID_L1_ADDRESS, *to_split_uint(10)]
        starknet.consume_message_from_l2(
            from_address=l2_bridge.contract_address,
            to_address=L1_BRIDGE_ADDRESS,
            payload=payload,
        )

    await check_balances(100, 100)

File: test/l2/l2_dai_bridge.ts
import { expect } from "chai";
import hre, { starknet } from "hardhat";
import { Account } from "@shardlabs/starknet-hardhat-plugin/dist/src/account";

import { eth, simpleDeployL2, MAX, strToFelt } from "../utils";
import { StarknetContract } from "@shardlabs/starknet-hardhat-plugin/dist/src/types";

import { unlinkSync } from "fs";

export async function snPredeployedAccounts2(n: number): Promise<Account[]> {
  const accounts: Account[] = [];

  for (const { address, private_key } of (
    await hre.starknet.devnet.getPredeployedAccounts()
  ).slice(0, n)) {
    accounts.push(
      await hre.starknet.OpenZeppelinAccount.getAccountFromAddress(
        address,
        private_key
      )
    );
  }
  return accounts;
}

describe("l2:bridge", async function () {
  this.timeout(900_000); // eslint-disable-line

  let l2Auth: Account;
  let bob: Account;
  let alice: Account;
  let foo: Account;

  let l2Dai: StarknetContract;
  let l2Bridge: StarknetContract;

  let l1BridgeAddress = "0x9F96fE0633eE838D0298E8b8980E6716bE81388d";
  const l1Recipient = "0x8aa7c51A6D380F4d9E273adD4298D913416031Ec";

  async function checkBalances(aliceBalance: bigint, bobBalance: bigint) {
    const { response: aliceBalanceOf } = await l2Dai.call("balance_of", {
      user: alice.address,
    });
    const { response: bobBalanceOf } = await l2Dai.call("balance_of", {
      user: bob.address,
    });
    expect(aliceBalanceOf).to.be.eq(aliceBalance);
    expect(bobBalanceOf).to.be.eq(bobBalance);
  }

  before(async function () {
    await hre.starknet.devnet.restart();

    [l2Auth, alice, bob, foo] = await snPredeployedAccounts2(4);

    l2Dai = await simpleDeployL2(
      l2Auth,
      "dai",
      {
        ward: l2Auth.address,
      },
      hre
    );

    l2Bridge = await simpleDeployL2(
      l2Auth,
      "l2_dai_bridge",
      {
        ward: l2Auth.address,
        dai: l2Dai.address,
        bridge: l1BridgeAddress,
      },
      hre
    );

    await l2Auth.invoke(l2Dai, "mint", {
      recipient: alice.address,
      amount: eth("100"),
    });

    await l2Auth.invoke(l2Dai, "mint", {
      recipient: bob.address,
      amount: eth("100"),
    });

    await l2Auth.invoke(l2Dai, "rely", {
      user: l2Bridge.address,
    });

    await hre.starknet.devnet.dump("./dump.json");
  });

  beforeEach(async () => {
    await hre.starknet.devnet.load("./dump.json");
  });

  after(async () => {
    try {
      unlinkSync("./dump.json");
    } catch (e) {}
  });

  it("initiate withdraw", async () => {
    await alice.invoke(l2Dai, "approve", {
      spender: l2Bridge.address,
      amount: eth("10"),
    });

    const txHash = await alice.invoke(l2Bridge, "initiate_withdraw", {
      l1_recipient: l1Recipient,
      amount: eth("10"),
    });
    const receipt = await starknet.getTransactionReceipt(txHash);
    const decodedEvents = l2Bridge.decodeEvents(receipt.events);

    expect(decodedEvents[0]).to.deep.eq({
      name: "WithdrawInitiated",
      data: {
        l1_recipient: BigInt(l1Recipient),
        amount: eth("10"),
        caller: BigInt(alice.address),
      },
    });
    // TODO: uncomment when payload type is fixed in the plugin
    // starknet.devnet.consumeMessageFromL2(
    //   l2Bridge.address,
    //   l1BridgeAddress,
    //   [0, l1Recipient, eth("10")]
    // )

    await checkBalances(eth("90"), eth("100"));
  });

  it("close should fail when not authorized", async () => {
    try {
      await alice.invoke(l2Bridge, "close");
    } catch (err: any) {
      expect(err.message).to.include(strToFelt("l2_dai_bridge/not-authorized"));
    }
  });

  it("initiate withdraw should fail when closed", async () => {
    await alice.invoke(l2Dai, "approve", {
      spender: l2Bridge.address,
      amount: eth("10"),
    });

    await l2Auth.invoke(l2Bridge, "close");

    const l1Recipient = "0x8aa7c51A6D380F4d9E273adD4298D913416031Ec";
    try {
      await alice.invoke(l2Bridge, "initiate_withdraw", {
        l1_recipient: l1Recipient,
        amount: eth("10"),
      });
    } catch (err: any) {
      expect(err.message).to.include(strToFelt("l2_dai_bridge/bridge-closed"));
    }
    await checkBalances(eth("100"), eth("100"));
  });

  it("initiate withdraw insufficient funds", async () => {
    const l1Recipient = "0x8aa7c51A6D380F4d9E273adD4298D913416031Ec";
    try {
      await foo.invoke(l2Bridge, "initiate_withdraw", {
        l1_recipient: l1Recipient,
        amount: eth("10"),
      });
    } catch (err: any) {
      expect(err.message).to.include(strToFelt("dai/insufficient-balance"));
    }
    await checkBalances(eth("100"), eth("100"));
  });

  it("withdraw invalid l1 address", async () => {
    const INVALID_L1_ADDRESS = 2n ** 251n;

    await alice.invoke(l2Dai, "approve", {
      spender: l2Bridge.address,
      amount: eth("10"),
    });

    try {
      await alice.invoke(l2Bridge, "initiate_withdraw", {
        l1_recipient: INVALID_L1_ADDRESS,
        amount: eth("10"),
      });
    } catch (err: any) {
      expect(err.message).to.include(
        strToFelt("Input too short for arguments")
      );
    }
    await checkBalances(eth("100"), eth("100"));
  });

  it("deposit", async () => {
    const { transaction_hash: txHash } = await starknet.devnet.sendMessageToL2(
      l2Bridge.address,
      "handle_deposit",
      l1BridgeAddress,
      [BigInt(bob.address), eth("10"), 0n, 0n],
      0,
      1n
    );
    const receipt = await starknet.getTransactionReceipt(txHash);
    const decodedEvents = l2Bridge.decodeEvents(receipt.events);

    expect(decodedEvents[0]).to.deep.eq({
      name: "DepositHandled",
      data: {
        account: BigInt(bob.address),
        amount: eth("10"),
      },
    });
    await checkBalances(eth("100"), eth("110"));
  });
});
